go/pools/smartconnpool/stack.go

```diff
@@ -0,0 +1,78 @@
+/*
+Copyright 2023 The Vitess Authors.
+
+Licensed under the Apache License, Version 2.0 (the "License");
+you may not use this file except in compliance with the License.
+You may obtain a copy of the License at
+
+    http://www.apache.org/licenses/LICENSE-2.0
+
+Unless required by applicable law or agreed to in writing, software
+distributed under the License is distributed on an "AS IS" BASIS,
+WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+See the License for the specific language governing permissions and
+limitations under the License.
+*/
+
+package smartconnpool
+
+import (
+	"runtime"
+	"sync/atomic"
+)
+
+// connStack is a lock-free stack for Connection objects. It is safe to
+// use from several goroutines.
+type connStack[C Connection] struct {
+	top atomic.Pointer[Pooled[C]]
+}
+
+func (s *connStack[C]) Push(item *Pooled[C]) {
+	for {
+		oldHead := s.top.Load()
+		item.next.Store(oldHead)
+		if s.top.CompareAndSwap(oldHead, item) {
+			return
+		}
+		runtime.Gosched()
```

Comment:
This is to avoid deadlocks?

Reply:
No, this yields the execution to another Goroutine to prevent "busy waiting". The Go runtime can no longer deadlock this way (since 1.8 IIRC) because it has a preemptive scheduler that can interrupt busy loops at any point, but when the system is under contention, and we fail to perform the atomic swap, it's more efficient to allow somebody else to attempt it instead of continuously trying for ourselves.

Reply:
Got it. Thank-you for the explanation!